### PR TITLE
KAFKA-12257:  Consumer mishandles topics deleted and recreated with the same name

### PR DESCRIPTION
**Pull request by jolshan**
Store topic ID info in consumer metadata. We will always take the topic ID from the latest metadata response and remove any topic IDs from the cache if the metadata response did not return a topic ID for the topic. 

With the addition of topic IDs, when we encounter a new topic ID (recreated topic) we can choose to get the topic's metadata even if the epoch is lower than the deleted topic.

The idea is that when we update from no topic IDs to using topic IDs, we will not count the topic as new (It could be the same topic but with a new ID). We will only take the update if the topic ID changed.

Added tests for this scenario as well as some tests for storing the topic IDs. Also added tests for topic IDs in metadata cache.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java [hunk 1]
```diff
@@ -93,6 +104,14 @@ private MetadataCache(String clusterId,
         return Optional.ofNullable(metadataByPartition.get(topicPartition));
     }
 
+    Map<String, Uuid> topicIds() {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/MetadataCache.java]: Do we need to expose the map or could we just have lookup methods:
```java
Uuid topicId(String topicName);
String topicName(Uuid topicId);
```

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java [hunk 2]
```diff
@@ -130,13 +150,30 @@ MetadataCache mergeWith(String newClusterId,
...
                             BiPredicate<String, Boolean> retainTopic) {
 
         Predicate<String> shouldRetainTopic = topic -> retainTopic.test(topic, internalTopics.contains(topic));
 
         Map<TopicPartition, PartitionMetadata> newMetadataByPartition = new HashMap<>(addPartitions.size());
+        Map<String, Uuid> newTopicIds = new HashMap<>(topicIds.size());
+
+        // We want the most recent topic ID. We add the old one here for retained topics and then update with newest information in the MetadataResponse
+        // we add if a new topic ID is added or remove if the request did not support topic IDs for this topic.
+        for (Map.Entry<String, Uuid> entry : this.topicIds.entrySet()) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/MetadataCache.java]: nit: map iterations are a little more readable with a `forEach`
```java
this.topicIds.forEach((topicName, topicId) -> {
  ...
});
```

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 3]
```diff
@@ -367,13 +384,14 @@ else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED)
...
-            boolean hasReliableLeaderEpoch) {
+            boolean hasReliableLeaderEpoch,
+            boolean changedTopicId) {
         TopicPartition tp = partitionMetadata.topicPartition;
         if (hasReliableLeaderEpoch && partitionMetadata.leaderEpoch.isPresent()) {
             int newEpoch = partitionMetadata.leaderEpoch.get();
             // If the received leader epoch is at least the same as the previous one, update the metadata
             Integer currentEpoch = lastSeenLeaderEpochs.get(tp);
-            if (currentEpoch == null || newEpoch >= currentEpoch) {
+            if (currentEpoch == null || newEpoch >= currentEpoch || changedTopicId) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: The log message below does not make much sense if the topicId has changed. Maybe this should be a separate case?

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 4]
```diff
@@ -316,8 +325,16 @@ private MetadataCache handleMetadataResponse(MetadataResponse metadataResponse,
         Set<String> invalidTopics = new HashSet<>();
 
         List<MetadataResponse.PartitionMetadata> partitions = new ArrayList<>();
+        Map<String, Uuid> topicIds = new HashMap<>();
         for (MetadataResponse.TopicMetadata metadata : metadataResponse.topicMetadata()) {
             topics.add(metadata.topic());
+            boolean changedTopicId = false;
+            if (!metadata.topicId().equals(Uuid.ZERO_UUID)) {
+                topicIds.put(metadata.topic(), metadata.topicId());
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: nit: maybe we can pull out a variable for `metadata.topic()` since there are 10 or so uses

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java [hunk 5]
```diff
@@ -130,13 +150,30 @@ MetadataCache mergeWith(String newClusterId,
...
+            }
+        }
+
         for (PartitionMetadata partition : addPartitions) {
             newMetadataByPartition.put(partition.topicPartition, partition);
+            Uuid id = topicIds.get(partition.topic());
+            if (id != null)
+                newTopicIds.put(partition.topic(), id);
+            else
+                // Remove if the latest metadata does not have a topic ID
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/MetadataCache.java:166]: What is the rationale to discard topicId information? Is this to deal with downgrades?

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 6]
```diff
@@ -874,18 +920,34 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
         assertEquals(cluster.topics(), new HashSet<>(Arrays.asList("keepValidTopic", "newValidTopic")));
         assertEquals(cluster.partitionsForTopic("keepValidTopic").size(), 2);
         assertEquals(cluster.partitionsForTopic("newValidTopic").size(), 4);
+        assertTrue(cluster.topicIds().containsAll(topicIds.values()));
+
+        // Try removing the topic ID from keepValidTopic (simulating receiving a request from a controller with an older IBP)
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: This test case is a bit much to take in. Is it possible to do a separate test case?

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java (hunk 5 above)
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/MetadataCache.java:166]: Yes, for the fetch path, we want to know when topic IDs are removed as quickly as possible so we can switch over to the older fetch version that uses topic names.

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java (hunk 1 above)
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/MetadataCache.java]: Hmm. I suppose we could have lookup methods. This has implications for the Fetch PR though.

Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/MetadataCache.java]: It is used when getting all the topic IDs to put into the fetch request/session in Fetcher there. Maybe it is ok to call a method multiple times there. I also use it in tests, but maybe we could change that usage.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 7]
```diff
@@ -391,10 +393,15 @@ else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED)
             int newEpoch = partitionMetadata.leaderEpoch.get();
             // If the received leader epoch is at least the same as the previous one, update the metadata
             Integer currentEpoch = lastSeenLeaderEpochs.get(tp);
-            if (currentEpoch == null || newEpoch >= currentEpoch || changedTopicId) {
+            if (currentEpoch == null || newEpoch >= currentEpoch) {
                 log.debug("Updating last seen epoch for partition {} from {} to epoch {} from new metadata", tp, currentEpoch, newEpoch);
                 lastSeenLeaderEpochs.put(tp, newEpoch);
                 return Optional.of(partitionMetadata);
+            } else if (changedTopicId) {
+                log.debug("Topic ID changed, so this topic must have been recreated. " +
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: nit: seems like it would be more useful for the log message to indicate the topic ids that changed instead of the unrelated epochs.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 8]
```diff
@@ -216,6 +217,14 @@ public synchronized boolean updateRequested() {
         }
     }
 
+    public synchronized Uuid topicId(String topicName) {
+        return cache.topicId(topicName);
+    }
+
+    public synchronized  String topicName(Uuid topicId) {
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: nit: 2 spaces between `synchronized` and `String`

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 9]
```diff
@@ -367,7 +386,8 @@ else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED)
      */
     private Optional<MetadataResponse.PartitionMetadata> updateLatestMetadata(
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:387]: We should also update the java doc to mention the TopicId changed case

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java [hunk 10]
```diff
@@ -130,13 +149,30 @@ MetadataCache mergeWith(String newClusterId,
...
+                            Map<String, Uuid> topicIds,
                             BiPredicate<String, Boolean> retainTopic) {
 
         Predicate<String> shouldRetainTopic = topic -> retainTopic.test(topic, internalTopics.contains(topic));
 
         Map<TopicPartition, PartitionMetadata> newMetadataByPartition = new HashMap<>(addPartitions.size());
+        Map<String, Uuid> newTopicIds = new HashMap<>(topicIds.size());
+
+        // We want the most recent topic ID. We add the old one here for retained topics and then update with newest information in the MetadataResponse
+        // we add if a new topic ID is added or remove if the request did not support topic IDs for this topic.
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/MetadataCache.java]: Can't get what you mean here. Could you rephrase this sentence?

**Review by showuon — COMMENTED** [review]
@jolshan , Thanks for the nice and simple fix! Overall LGTM! Left some minor comments. Thanks.

**Comment by showuon** [comment]
Also, please remember to rebase with the latest `trunk` branch. Thanks.

**Comment by jolshan** [comment]
Yup. This merge conflict was caused by my other PR 😅

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 11]
```diff
@@ -377,6 +397,12 @@ else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED)
                 log.debug("Updating last seen epoch for partition {} from {} to epoch {} from new metadata", tp, currentEpoch, newEpoch);
                 lastSeenLeaderEpochs.put(tp, newEpoch);
                 return Optional.of(partitionMetadata);
+            // If the topic ID changed, updated the metadata
+            } else if (changedTopicId) {
+                log.debug("Topic ID changed, so this topic must have been recreated. " +
+                        "Removing last seen epoch {} for the old partition {} and adding epoch {} from new metadata", currentEpoch, tp, newEpoch);
+                lastSeenLeaderEpochs.put(tp, newEpoch);
+                return Optional.of(partitionMetadata);
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:406]: Leaving this comment here for lack of an alternative location. This patch takes a good first step in improving consumer behavior for the topic recreation case. At least we are able to detect and discard the old epoch state. In fact, it does a little more than that since, combined with the fetch validation logic, we are likely to detect that the old fetch position is no longer valid. Most likely this case would get raised to the user as a `LogTruncationException`, which might not be ideal, but at least is justifiable. However, it doesn't quite close the door on reuse of the fetch position since it may remain valid on the recreated topic. For the full solution, we probably need to track topicId in SubscriptionState as well so that we can force an offset reset whenever the topicId changes. I think it makes sense to do this in https://issues.apache.org/jira/browse/KAFKA-12975.

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java [hunk 12]
```diff
@@ -93,6 +104,13 @@ private MetadataCache(String clusterId,
         return Optional.ofNullable(metadataByPartition.get(topicPartition));
     }
 
+    Uuid topicId(String topicName) {
+        return topicIds.get(topicName);
+    }
+    String topicName(Uuid topicId) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/MetadataCache.java]: nit: add newline between methods

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java (hunk 5 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/MetadataCache.java:166]: Does this still make sense in the context of 3.0, which does not have topicId fetch logic?

Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/MetadataCache.java:166]: I suppose it is not needed, but I'm not sure if it helps a lot to remove it.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 11 above)
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java:406]: I see. I think the main issue here was that we would ignore metadata updates when we were simply looking at the epoch. I believe that this PR solves the problem, but we can continue to improve beyond this.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 7 above)
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Hmm I don't actually have access to both the topic IDs in this method. I can just do the old one and/or pass in the parameter for the new one.

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java]: If you pass the new one, then you can probably get rid of `changedTopicId`

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 11 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java:406]: Yes, I was just pointing out that there is still a gap.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 7 above)
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Yeah, I was thinking that too. I just have to be careful when comparing to remember the zero uuid case.

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java (hunk 5 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/MetadataCache.java:166]: We can leave it as is I guess since I can't think of a strong case to remove it. It is a rare situation that we would hit this case and the consequence of losing the topic ID is probably not too bad. Worst case, we might miss a recreation which occurred while the cluster was rolling to upgrade or downgrade. On the other hand, it could lead to other kinds of problems if we allow updates to the epoch information tied to a topic ID without being able to validate that the topic ID is correct, so maybe this logic is for the best.

Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/MetadataCache.java:166]: Yeah. That was my reasoning. I thought the upgrade/downgrade case would be rare and the guarantees harder to reason about there.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 13]
```diff
@@ -377,6 +397,12 @@ else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED)
                 log.debug("Updating last seen epoch for partition {} from {} to epoch {} from new metadata", tp, currentEpoch, newEpoch);
                 lastSeenLeaderEpochs.put(tp, newEpoch);
                 return Optional.of(partitionMetadata);
+            // If both topic IDs were valid and the topic ID changed, update the metadata
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: nit: move this comment into the `if`

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 14]
```diff
@@ -377,6 +397,12 @@ else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED)
...
+            } else if (!topicId.equals(Uuid.ZERO_UUID) && oldTopicId != null && !topicId.equals(oldTopicId)) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Hmm, shouldn't this check come before the epoch check? Admittedly, it's unlikely that a recreated topic would have a higher epoch, but we may as well handle that case.

By the way, it's a little inconsistent that this check uses both null and `Uuid.ZERO_UUID` to represent a missing value. Maybe we can use null consistently?

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 15]
```diff
@@ -316,20 +325,29 @@ private MetadataCache handleMetadataResponse(MetadataResponse metadataResponse,
         Set<String> invalidTopics = new HashSet<>();
 
         List<MetadataResponse.PartitionMetadata> partitions = new ArrayList<>();
+        Map<String, Uuid> topicIds = new HashMap<>();
         for (MetadataResponse.TopicMetadata metadata : metadataResponse.topicMetadata()) {
-            topics.add(metadata.topic());
+            String topicName = metadata.topic();
+            Uuid topicId = metadata.topicId();
+            topics.add(topicName);
+            // We only update if the current metadata since we can only compare when both topics have valid IDs
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: nit: rephrase?

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 16]
```diff
@@ -316,20 +325,29 @@ private MetadataCache handleMetadataResponse(MetadataResponse metadataResponse,
...
+            Uuid oldTopicId = null;
+            if (!topicId.equals(Uuid.ZERO_UUID)) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: nit: I think we're guaranteed that `topicId` is not null (in spite of the inconsistent `equals`), but it's still a little nicer to write this check as `!Uuid.ZERO_UUID.equals(topicId)`

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 17]
```diff
@@ -216,6 +217,14 @@ public synchronized boolean updateRequested() {
         }
     }
 
+    public synchronized Uuid topicId(String topicName) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:223]: Can you document that this returns null if the topicId does not exist or is not known? Similarly for `topicName`.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 18]
```diff
@@ -377,6 +397,12 @@ else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED)
                 log.debug("Updating last seen epoch for partition {} from {} to epoch {} from new metadata", tp, currentEpoch, newEpoch);
                 lastSeenLeaderEpochs.put(tp, newEpoch);
                 return Optional.of(partitionMetadata);
+            // If both topic IDs were valid and the topic ID changed, update the metadata
+            } else if (!topicId.equals(Uuid.ZERO_UUID) && oldTopicId != null && !topicId.equals(oldTopicId)) {
+                log.debug("Topic ID for partition {} changed from {} to {}, so this topic must have been recreated. " +
+                                "Using the newly updated metadata.", tp, oldTopicId, topicId);
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Instead of "Using the newly updated metadata," maybe we can say this:
> Resetting the last seen epoch to {}.

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java [hunk 19]
```diff
@@ -69,13 +73,20 @@ private MetadataCache(String clusterId,
...
                           Cluster clusterInstance) {
         this.clusterId = clusterId;
         this.nodes = nodes;
         this.unauthorizedTopics = unauthorizedTopics;
         this.invalidTopics = invalidTopics;
         this.internalTopics = internalTopics;
         this.controller = controller;
+        this.topicIds = topicIds;
+
+        this.topicNames = new HashMap<>(topicIds.size());
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/MetadataCache.java]: As far as I can tell, there are no uses of this collection (`Metadata.topicName` is not referenced). Can we get rid of it?

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java [hunk 20]
```diff
@@ -130,13 +150,36 @@ MetadataCache mergeWith(String newClusterId,
...
+                            Map<String, Uuid> topicIds,
                             BiPredicate<String, Boolean> retainTopic) {
 
         Predicate<String> shouldRetainTopic = topic -> retainTopic.test(topic, internalTopics.contains(topic));
 
         Map<TopicPartition, PartitionMetadata> newMetadataByPartition = new HashMap<>(addPartitions.size());
+        Map<String, Uuid> newTopicIds = new HashMap<>(topicIds.size());
+
+        // We want the most recent topic ID. We start with the previous ID stored for retained topics and then
+        // update with newest information in the MetadataResponse.
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/MetadataCache.java]: nit: "update with the newest information from the MetadataResponse."

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java [hunk 21]
```diff
@@ -130,13 +150,36 @@ MetadataCache mergeWith(String newClusterId,
...
+        // If the newest MetadataResponse:
+        //    - contains a new topic with no ID, add no IDs to newTopicIds
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/MetadataCache.java]: nit: some of these cases do not seem worth mentioning. I think we're just saying that we always take the latest state, removing existing topicIds if necessary.

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 22]
```diff
@@ -371,6 +372,43 @@ public void testUpdateLastEpoch() {
...
+        metadataResponse = RequestTestUtils.metadataUpdateWithIds("dummy", 1, Collections.emptyMap(), Collections.singletonMap("topic-1", 1), _tp -> 10, topicIds);
+        metadata.updateWithCurrentRequestVersion(metadataResponse, false, 3L);
+        assertOptional(metadata.lastSeenLeaderEpoch(tp), leaderAndEpoch -> assertEquals(leaderAndEpoch.intValue(), 10));
+
+        // Update if we see newer epoch
+        metadataResponse = RequestTestUtils.metadataUpdateWithIds("dummy", 1, Collections.emptyMap(), Collections.singletonMap("topic-1", 1), _tp -> 12, topicIds);
+        metadata.updateWithCurrentRequestVersion(metadataResponse, false, 4L);
+        assertOptional(metadata.lastSeenLeaderEpoch(tp), leaderAndEpoch -> assertEquals(leaderAndEpoch.intValue(), 12));
+
+        // We should also update if we see a new topicId even if the epoch is lower
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:411]: We may as well cover the case when the topicId is changed _and_ the epoch is higher.

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 23]
```diff
@@ -371,6 +372,43 @@ public void testUpdateLastEpoch() {
...
+        // Start with a topic with no topic ID
+        metadataResponse = RequestTestUtils.metadataUpdateWith("dummy", 1, Collections.emptyMap(), Collections.singletonMap("topic-1", 1), _tp -> 10);
+        metadata.updateWithCurrentRequestVersion(metadataResponse, false, 1L);
+        assertOptional(metadata.lastSeenLeaderEpoch(tp), leaderAndEpoch -> assertEquals(leaderAndEpoch.intValue(), 10));
+
+        // We should treat an added topic ID as though it is the same topic. Handle only when epoch increases.
+        // Don't update to an older one
+        metadataResponse = RequestTestUtils.metadataUpdateWithIds("dummy", 1, Collections.emptyMap(), Collections.singletonMap("topic-1", 1), _tp -> 1, topicIds);
+        metadata.updateWithCurrentRequestVersion(metadataResponse, false, 2L);
+        assertOptional(metadata.lastSeenLeaderEpoch(tp), leaderAndEpoch -> assertEquals(leaderAndEpoch.intValue(), 10));
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: nit: this seems more concise
```java
        assertEquals(Optional.of(10), metadata.lastSeenLeaderEpoch(tp));
```

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 24]
```diff
@@ -863,8 +906,12 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
             "keepValidTopic",
             "newValidTopic")));
 
-        metadataResponse = RequestTestUtils.metadataUpdateWith(newClusterId, newNodes, newTopicErrors, newTopicPartitionCounts, _tp -> 200);
+        topicIds.put("newValidTopic", Uuid.randomUuid());
+        metadataResponse = RequestTestUtils.metadataUpdateWithIds(newClusterId, newNodes, newTopicErrors, newTopicPartitionCounts, _tp -> 200, topicIds);
         metadata.updateWithCurrentRequestVersion(metadataResponse, true, time.milliseconds());
+        topicIds.remove("oldValidTopic");
+        retainTopics.get().forEach(topic -> assertEquals(metadata.topicId(topic), topicIds.get(topic)));
+        assertEquals(metadata.topicId("oldValidTopic"), null);
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: nit: use `assertNull`

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 25]
```diff
@@ -845,6 +887,7 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
         assertEquals(cluster.topics(), new HashSet<>(Arrays.asList("oldValidTopic", "keepValidTopic")));
         assertEquals(cluster.partitionsForTopic("oldValidTopic").size(), 2);
         assertEquals(cluster.partitionsForTopic("keepValidTopic").size(), 3);
+        assertTrue(cluster.topicIds().containsAll(topicIds.values()));
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: Do we want to make this assertion stronger? Or is `topicIds` a subset?

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 26]
```diff
@@ -874,18 +921,69 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
...
         assertEquals(cluster.partitionsForTopic("newValidTopic").size(), 4);
+        assertTrue(cluster.topicIds().containsAll(topicIds.values()));
 
         // Perform another metadata update, but this time all topic metadata should be cleared.
         retainTopics.set(Collections.emptySet());
 
-        metadataResponse = RequestTestUtils.metadataUpdateWith(newClusterId, newNodes, newTopicErrors, newTopicPartitionCounts, _tp -> 300);
+        metadataResponse = RequestTestUtils.metadataUpdateWithIds(newClusterId, newNodes, newTopicErrors, newTopicPartitionCounts, _tp -> 300, topicIds);
         metadata.updateWithCurrentRequestVersion(metadataResponse, true, time.milliseconds());
+        topicIds.forEach((topicName, topicId) -> assertEquals(metadata.topicId(topicName), null));
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: nit: `assertNull`

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 27]
```diff
@@ -874,18 +921,69 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
...
+        };
+
+        // Initialize a metadata instance with two topics. Both will be retained.
+        String clusterId = "clusterId";
+        int nodes = 2;
+        Map<String, Integer> topicPartitionCounts = new HashMap<>();
+        topicPartitionCounts.put("validTopic1", 2);
+        topicPartitionCounts.put("validTopic2", 3);
+
+        retainTopics.set(new HashSet<>(Arrays.asList(
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: nit: you can use `Utils.mkSet` (a few more of these)

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 28]
```diff
@@ -874,18 +921,69 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
...
+
+        Cluster cluster = metadata.fetch();
+        // We still have the topic, but it just doesn't have an ID.
+        assertEquals(cluster.topics(), new HashSet<>(Arrays.asList("validTopic1", "validTopic2")));
+        assertEquals(cluster.partitionsForTopic("validTopic1").size(), 2);
+        assertTrue(cluster.topicIds().containsAll(topicIds.values()));
+        assertEquals(Uuid.ZERO_UUID, cluster.topicId("validTopic1"));
+
     }
+
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:999]: nit: extra newline (and above before the end brace)

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 29]
```diff
@@ -874,18 +921,69 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
...
+        topicIds.remove("validTopic1");
+        metadataResponse = RequestTestUtils.metadataUpdateWithIds(clusterId, nodes, Collections.emptyMap(), topicPartitionCounts, _tp -> 200, topicIds);
+        metadata.updateWithCurrentRequestVersion(metadataResponse, true, time.milliseconds());
+        retainTopics.get().forEach(topic -> assertEquals(metadata.topicId(topic), topicIds.get(topic)));
+
+        Cluster cluster = metadata.fetch();
+        // We still have the topic, but it just doesn't have an ID.
+        assertEquals(cluster.topics(), new HashSet<>(Arrays.asList("validTopic1", "validTopic2")));
+        assertEquals(cluster.partitionsForTopic("validTopic1").size(), 2);
+        assertTrue(cluster.topicIds().containsAll(topicIds.values()));
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: nit: I think we can make this assertion stronger

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 30]
```diff
@@ -874,18 +921,69 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
...
+
+        // Try removing the topic ID from keepValidTopic (simulating receiving a request from a controller with an older IBP)
+        topicIds.remove("validTopic1");
+        metadataResponse = RequestTestUtils.metadataUpdateWithIds(clusterId, nodes, Collections.emptyMap(), topicPartitionCounts, _tp -> 200, topicIds);
+        metadata.updateWithCurrentRequestVersion(metadataResponse, true, time.milliseconds());
+        retainTopics.get().forEach(topic -> assertEquals(metadata.topicId(topic), topicIds.get(topic)));
+
+        Cluster cluster = metadata.fetch();
+        // We still have the topic, but it just doesn't have an ID.
+        assertEquals(cluster.topics(), new HashSet<>(Arrays.asList("validTopic1", "validTopic2")));
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: nit: the first argument should be the expected one

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 14 above)
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java]: This bugged me a bit too. The issue is that the request itself uses Uuid.ZERO_UUID, so we'd just have to convert that to null. We can do that if it is clearer to read.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 16 above)
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Yeah. There can not be a null uuid in the request. but I can rephrase

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 17 above)
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java:223]: Will do

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 18 above)
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java]: So part of the reason why I say this is that this method decides whether we use or ignore the new metadata to update the cache in general. But I can change to the epoch message.

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java (hunk 19 above)
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/MetadataCache.java]: sure

File: clients/src/main/java/org/apache/kafka/clients/MetadataCache.java (hunk 21 above)
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/MetadataCache.java]: sure. that's easier to understand I think.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 14 above)
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Also, these checks result in the same thing. The only difference is the log debug line. If it makes more sense to log the topic ID change, I can switch the order.

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java (hunk 25 above)
Comment (jolshan) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: To make it stronger, do we just also assert the inverse `(topicIds.values().containsAll(cluster.topicIds())` Or am I being silly and we can just check equality here?

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java (hunk 30 above)
Comment (jolshan) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: I can switch it, but I was trying to keep the ordering consistent in the file. For example. The test above does the same thing:
`assertEquals(cluster.topics(), new HashSet<>(Arrays.asList("keepValidTopic", "newValidTopic")));`

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java (hunk 23 above)
Comment (jolshan) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: Should I update this for the test above as well?

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java (hunk 25 above)
Comment (jolshan) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: The collection types are different so I made a helper method to test both containAll. There may have been a helper somewhere else in the code but I couldn't find it.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 14 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Yes, the logging is what I had in mind. The log message is misleading otherwise.

Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Ah ok. I'll switch it.

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the patch!
